### PR TITLE
Bump LibLUV to 1.34.1

### PR DESCRIPTION
**Pull request by lewis6991**

### CODE REVIEW

**Comment by clason** [comment]
Release notes: https://github.com/luvit/luv/releases/tag/1.34.1-0
(Introduces `uv.fs_mkstemp`, which is the point here, I believe?)

Note that 1.34.2+ introduce breaking changes to the CMake build, so that's the latest version that's safe to update. (They also introduce useful changes that would make it worth adapting the build scripts -- in a future PR.)

**Comment by lewis6991** [comment]
Yep. For some reason null-ls.nvim is using `fs_mkstemp` even though it isn't exposed. I've opened a PR on the plugin to make it work with 0.5: https://github.com/jose-elias-alvarez/null-ls.nvim/pull/23

**Comment by clason** [comment]
They're probably using a distribution build -- homebrew links against 1.41.0 

(So we know it's safe to update once we get the build system worked out :P)

**Comment by clason** [comment]
I just looked at the actual commit: I don't think that's enough -- you also need to change the version in https://github.com/neovim/neovim/blob/18bfcfe77f007f146a7251aa3cc6129891390c30/third-party/CMakeLists.txt#L178

(and adapt the hash below).

**Comment by lewis6991** [comment]
Many thanks. As you can see... I barely know what I'm doing with all this cmake business 😜

**Comment by clason** [comment]
Neither do I, except for what I have learned by painful trial and error (and help from jamessan) ;)

**Comment by clason** [comment]
Hmm. Looks like this needs some other dependency bumps (`lua-compat-53`?) The build passed locally for me, though.

**Comment by lewis6991** [comment]
Ok this passes for me locally from a clean checkout.

**Comment by clason** [comment]
Same here. LGTM -- @jamessan ?

File: CMakeLists.txt [hunk 1]
```diff
@@ -387,7 +387,7 @@ include_directories(SYSTEM ${LIBUV_INCLUDE_DIRS})
 find_package(Msgpack 1.0.0 REQUIRED)
 include_directories(SYSTEM ${MSGPACK_INCLUDE_DIRS})
 
-find_package(LibLUV 1.30.0 REQUIRED)
+find_package(LibLUV 1.34.1 REQUIRED)
```
[the remark this answers is not in the document]
Comment (lewis6991) [reply CMakeLists.txt:390]: For plugins that want to use more exposed functions. See: https://github.com/jose-elias-alvarez/null-ls.nvim/pull/23

[the remark this answers is not in the document]
Comment (clason) [reply CMakeLists.txt:390]: History: That was the first try to bump the bundled version :)

(We're not using it, but plugins are, and we might? `uv.fs_mkstem`)

[the remark this answers is not in the document]
Comment (jamessan) [reply CMakeLists.txt:390]: > For plugins that want to use more exposed functions. See: https://github.com/jose-elias-alvarez/null-ls.nvim/pull/23

Some plugins also require ffi, but that doesn't mean nvim does. This specifies the minimum that nvim requires.

Plugins should handle validating their base requirements are satisfied or use alternatives, rather than assuming nvim is built against a certain dependency version.

[the remark this answers is not in the document]
Comment (clason) [reply CMakeLists.txt:390]: Makes sense -- we haven't done that so far, and we shouldn't start now (with a single dependency).

_If_ that is done, it should be done across the board and documented as a breaking change. (The bump to `third-party` is sufficient to make the default build use the newer version, so leaving the minimum version lower is fine.)

[the remark this answers is not in the document]
Comment (lewis6991) [reply CMakeLists.txt:390]: > (The bump to third-party is sufficient to make the default build use the newer version, so leaving the minimum version lower is fine.)

I don't follow. ELI5? Is there a min version and a version for using more up-to-date dependencies?

[the remark this answers is not in the document]
Comment (clason) [reply CMakeLists.txt]: Sorry.

So the `Lib_LUV` in the _toplevel_ `CMakeLists.txt` is the minimum version required for Neovim itself to build successfully -- it tells distributions not to try to build against _earlier_ versions by erroring out early, because the basic functionality is known to be broken then.

The version in `third-party/CMakeLists.txt` is the version Neovim bundles in CI and for the _official_ releases -- this can be include nice-to-have features but is usually ignored by distributions (who build against their own versions, which is fine as long as that version is newer than the one specified in the top-level file). 

So the latter is what you want to make newer features accessible to the plugin ecosystem (in order not to fall behind distributions); the former is what you want to _force_ distributions to upgrade their own versions (which you'd only do if Neovim itself is unusable with earlier versions).

[the remark this answers is not in the document]
Comment (lewis6991) [reply CMakeLists.txt]: Got it. Thanks for explaining.

[the remark this answers is not in the document]
Comment (clason) [reply CMakeLists.txt]: (Teaching = learning + sharing :P)

**Comment by lewis6991** [comment]
CI/OSX has failed but I suspect it's flaky CI since the failure is with oldtests. LibLUV shouldn't effect those right?

```
2021-07-04T19:15:30.8908380Z Executed:  2072 Tests
2021-07-04T19:15:30.8908730Z  Skipped:   183 Tests
2021-07-04T19:15:30.8909090Z   FAILED:     1 Tests
2021-07-04T19:15:30.8909310Z 
2021-07-04T19:15:30.8909490Z 
2021-07-04T19:15:30.8909780Z Failures: 
2021-07-04T19:15:30.8910130Z 	From test_autocmd.vim:
2021-07-04T19:15:30.8910610Z 	Found errors in Test_cursorhold_insert():
2021-07-04T19:15:30.8910990Z 	Run 1:
2021-07-04T19:15:30.8911500Z 	function RunTheTest[37]..Test_cursorhold_insert line 9: Expected 1 but got 0
2021-07-04T19:15:30.8912000Z 	Run 2:
2021-07-04T19:15:30.8912500Z 	function RunTheTest[37]..Test_cursorhold_insert line 9: Expected 1 but got 0
2021-07-04T19:15:30.8913100Z 	Flaky test failed too often, giving up
```

**Comment by 0x777** [comment]
> Note that 1.34.2+ introduce breaking changes to the CMake build, so that's the latest version that's safe to update. (They also introduce useful changes that would make it worth adapting the build scripts -- in a future PR.)

@clason I opened a [PR](https://github.com/neovim/neovim/pull/14167) a while ago with changes required for 1.34.2+. The Makefile changes work on my linux machine but they fail on the CI's windows environment.

**Comment by clason** [comment]
@0x777 Thanks, I'll take a look at it! Depending on which version you tested, 1.41 at least bumps minimal support to Windows 8, so that would definitely be breaking. (But we also special-case the Windows version anyway, so similar case distinction may be necessary for the script changes.)

**Comment by clason** [comment]
@lewis6991 can't reproduce _that_ test failure locally on macOS, either. Try force-pushing (and possibly rebasing on master) again; let's see if that changes things.

**Comment by clason** [comment]
Now it's a functional test that fails 😆 

That test passes locally for me on macOS, though.

**Comment by lewis6991** [comment]
Different failure now, flakiness confirmed.

**Comment by clason** [comment]
And now it's all green :)
